lib/compiler/src/beam_jump.erl

```diff
@@ -383,6 +399,47 @@ insert_fc_labels([{label,L}=I|Is0], Mlbl) ->
     end;
 insert_fc_labels([_|_]=Is, _) -> Is.
 
+%% After eliminating a test, it might happen, that a register was only used
+%% in this test. Let's check if that was the case and if it was so, we can
+%% eliminate the load into the register completely.
+opt_useless_loads([{block,_}|_]=Is, Args, L, #st{index={lazy,FIs}}=St) ->
+    opt_useless_loads(Is, Args, L, St#st{index=beam_utils:index_labels(FIs)});
+opt_useless_loads([{block,Block0}|Is], Args, L, #st{index=Index}=St) ->
+    case opt_useless_block_loads(Block0, Args, L, Index) of
+        [] ->
+            opt_useless_loads(Is, Args, L, St);
+        [_|_]=Block ->
+            {[{block,Block}|Is],St}
+    end;
+%% After eliminating the test and useless blocks, it might happen,
+%% that the previous test could also be eliminated.
+%% It might be that the label was already marked as used, even if ultimately,
+%% it never will be - we can't do much about it at that point, though
+opt_useless_loads([{test,_,{f,L},Args1}=I|Is], Args, L, St) ->
+    case beam_utils:is_pure_test(I) of
+        false ->
+            {[I|Is],St};
+        true ->
+            opt_useless_loads(Is, Args ++ Args1, L, St)
+    end;
+opt_useless_loads(Is, _Args, _L, St) ->
+    {Is,St}.
+
+opt_useless_block_loads([{set,[Dst],_,_}=I|Is], Args, L, Index) ->
+    case lists:member(Dst, Args) andalso
+        (Is =:= [] orelse beam_utils:is_killed_block(Dst, Is)) andalso
```

Comment:
@bjorng I'm not sure why `beam_utils:is_killed_block` returns `false` for an empty block.